rfc/0058-error-handling.md

```diff
@@ -317,6 +317,73 @@ A Note on IDs: The IDs in this RFC are only for organisational purposes and MUST
 
 * Raised when decoding of the data into the user object failed
 
+### 21 RateLimitingFailure
+
+* Raised when a service decides that the caller must be rate limited due to exceeding a rate threshold of some sort.
+* Note that since there are many different reasons why a request is rate limited, the error context MUST include the reason / specific type of rate limiting cause for debugging purposes.
+
+Since the failure itself is a very generic error, an additional Set of `RateLimitingReason` should be part of the error that allows the user - if needed - to further inspect/narrow down the cause. Note that it is a set, since it could contain multiple reasons at the same time.
+
+```
+enum RateLimitingReason {
+  NetworkIngressRateLimitReached,
+  NetworkEgressRateLimitReached,
+  MaximumConnectionsReached,
+  RequestRateLimitReached,
+  ConcurrentRequestLimitReached
```

Comment:
I'm not sure about MaximumConnectionsReached being in RateLimitingFailure.  If it's a pooled service that's maybe accurate, but for KV it's going to be kvEndpoints and/or too many apps provoking this, which will require manual intervention and so would ideally be a QuotaLimitingFailure.
We could return RateLimitingFailure or QuotaLimitingFailure depending on whether it's a dynamic/pooled service or not.  But perhaps that's getting too sophisticated?

Reply:
@programmatix as you said it depends on the service and also how the user has set up their infra (you could imagine spinning up app servers on demand in k8s) vs. having them static. I'd vote for what we have is good enough, and if in doubt we can always bring it in with the other exception in the future as well.